### PR TITLE
shell: password strength indication

### PR DESCRIPTION
**Pull request by fridex**
Based on https://github.com/cockpit-project/cockpit/issues/2031 and https://github.com/cockpit-project/cockpit/issues/644

### CODE REVIEW

**Comment by stefwalter** [comment]
Nice work.

Could you rebase this ontop of git master. In particular, we've dropped the dependency on 'expect' once packagers informed us that it was implemented in TCL.

**Comment by fridex** [comment]
> Could you rebase this ontop of git master.

Yes, sure. It should be rebased by now.

**Comment by fridex** [comment]
Now I see, bad rebase. Sorry for that....

**Comment by andreasn** [comment]
Good start!
0% is hard to read due to white text on gray background and due to being glued to the left side of the progress bar.

screenshot from 2015-04-01 11 50 05

**Comment by andreasn** [comment]
The dialog will also need to inform that the current password is too weak to proceed and ideally give some tips on how you can strengthen it.

**Comment by andreasn** [comment]
Having it go by percentage from 0%-100% gives the notion that only 100% is acceptable. Does the backend have any notion of levels? Like: Bad password, Okay password, Good password, Excellent password? Similar to what Dropbox and the Fedora installer among others do http://blog.codinghorror.com/content/images/2015/01/dropbox-password-strength-meters.png

**Comment by andreasn** [comment]
Mockup

**Comment by fridex** [comment]
> The dialog will also need to inform that the current password is too weak to proceed and ideally give some tips on how you can strengthen it.

Tips could be confusing since it depends on actual configuration.

> Having it go by percentage from 0%-100% gives the notion that only 100% is acceptable. Does the backend have any notion of levels? Like: Bad password, Okay password, Good password, Excellent password?

There is only 0-100 scale (more than 0 is acceptable), but levels can be easily hard-coded.

According to the design, I suppose that progress bar is going to be replaced with something else. Is it already available in patternfly? .

**Comment by andreasn** [comment]
> Tips could be confusing since it depends on actual configuration.

All right. Does the system have any kind of clue if it's too short or not, or do we need to go for something generic like "Password is not strong enough" ?

> There is only 0-100 scale (more than 0 is acceptable), but levels can be easily hard-coded.

Cool. Lets try and do it in steps of 25%. 0-25%, 26-50%, 51-75%, 76-100%

> According to the design, I suppose that progress bar is going to be replaced with something else. Is it already available in patternfly? .

No, unfortunately not, but I'm happy to help with the css.

**Comment by fridex** [comment]
> All right. Does the system have any kind of clue if it's too short or not, or do we need to go for something generic like "Password is not strong enough" ?

Now I can imagine only a generic solution like yours.

> No, unfortunately not, but I'm happy to help with the css.

That would be great.

**Comment by andreasn** [comment]
so using this markup https://github.com/andreasn/cockpit/commit/a8a73999f41951e0b3b2f48cd4f1bce6a12b8ab6#diff-05e1d4cff6d7b149dc269fa20af3ebd5R201

and adding any of the classes weak/okay/good/excellent should get the desired results.
Not having any of those classes will give an empty bar, and that should be the initial state.

**Comment by andreasn** [comment]
Great progress!
Would it be possible to place the strengh meter in the same table as the password field, instead of in it's own table row?

screenshot from 2015-04-13 19 15 03

**Comment by andreasn** [comment]
In order to achieve "Excellent password" I need to type a couple of more characters after the bar has been filled. Would be better to have more of a 1-to-1 match between filled bar and Excellent password.

**Comment by fridex** [comment]
In order to achieve "excellent password", one has to enter password with 100% score. If it is desired to have 4 bars only when "excellent password" is entered, the scale should be adjusted. Would be 0-33, 34-66, 67-99, 100 OK? Or is it better to achieve "excellent password" with score more than 75%? Personally, I would prefer the first option.

**Comment by stefwalter** [comment]
> In order to achieve "excellent password", one has to enter password with 100% score. If it is desired to have 4 bars only when "excellent password" is entered, the scale should be adjusted. Would be 0-33, 34-66, 67-99, 100 OK?

Makes sense to me.

**Comment by andreasn** [comment]
Some layout fixes, as promised earlier: https://github.com/andreasn/cockpit/tree/strength-meter-layout

**Comment by fridex** [comment]
Oh, sorry. Looks like we are working on it at the same time :) I will use your patch.

**Comment by andreasn** [comment]
It seems this needs a rebase

**Comment by fridex** [comment]
Rebased.

**Comment by andreasn** [comment]
Looks good!

File: pkg/shell/shell.html [hunk 1]
```diff
@@ -1199,18 +1205,26 @@ <h4 class="modal-title">Set Password</h4>
...
+                    </div>
                   </div>
                 </td>
               </tr>
               <tr>
                 <td></td>
-                <td><div id="account-set-password-message-password-mismatch" class="has-error">
-                    <span class="help-block">The passwords do not match</span>
+                <td><div class="has-error">
+                    <span id="account-set-password-meter-message" class="help-block">
```
Comment (stefwalter) [inline pkg/shell/shell.html]: This needs ```translatable="yes"```

File: pkg/shell/shell.html [hunk 2]
```diff
@@ -1199,18 +1205,26 @@ <h4 class="modal-title">Set Password</h4>
                 </td>
               </tr>
               <tr>
-                <td translatable="yes">Confirm New Password</td>
+                <td translatable="yes" class="top">Confirm New Password</td>
```
Comment (stefwalter) [inline pkg/shell/shell.html]: This needs ```translatable="yes"```

File: pkg/shell/shell.html [hunk 3]
```diff
@@ -1129,19 +1129,25 @@ <h4 class="modal-title">Create New Account</h4>
...
                   </div>
                 </td>
               </tr>
               <tr>
                 <td></td>
                 <td>
-                  <div id="accounts-create-message-password-mismatch" class="has-error">
-                    <span class="help-block">The passwords do not match</span>
+                  <div class="has-error">
+                    <span id="accounts-create-password-meter-message" class="help-block">The passwords do not match</span>
```
Comment (stefwalter) [inline pkg/shell/shell.html]: This needs ```translatable="yes"```

File: pkg/shell/shell.html [hunk 4]
```diff
@@ -1129,19 +1129,25 @@ <h4 class="modal-title">Create New Account</h4>
                 </td>
               </tr>
               <tr>
-                <td translatable="yes">Confirm</td>
+                <td translatable="yes" class="top">Confirm</td>
```
Comment (stefwalter) [inline pkg/shell/shell.html]: This needs ```translatable="yes"```

**Comment by stefwalter** [comment]
Looks really good. I'll fix the remaining issues when merging.

**Comment by stefwalter** [comment]
The integration tests fail due to password validation. Since I have to fix things anyway. I'm going to migrate this to the new validation style. https://github.com/cockpit-project/cockpit/wiki/HIG:-Dialogs

**Comment by stefwalter** [comment]
In addition the validation doesn't actually work yet. It's easy to get into a state where it'll never allow you to proceed no matter how much you change the password.

**Comment by stefwalter** [comment]
1. by just typing and tabbing around, i can get things into a state where i just have to cancel out, i never get the 'Set' button enabled
  2. the logic doesn't match the HIG for dialogs, which is mostly because it didn't work that way before you started: https://github.com/cockpit-project/cockpit/wiki/HIG:-Dialogs
  3. lastly the check-accounts integration tests fails because they need to be updated for the new validation

**Comment by andreasn** [comment]
> 2. the logic doesn't match the HIG for dialogs
With regards to the insensitive button and stuff? Could we do that as a followup to this issue? It feels like we got in the middle of it all and outside the control of this pull request.

Issue one sounds more severe though.

**Comment by fridex** [comment]
> 1. by just typing and tabbing around, i can get things into a state where i just have to cancel out, i never get the 'Set' button enabled

Cannot reproduce. Could you be please more specific? Tab is working as expected on Chrome for me.

**Comment by stefwalter** [comment]
Rebased, squashed, fixed tests, updated dialog behavior in #2233

**Pull request closed without merging**